StandardFiniteFieldsOSCAR.jl

```diff
@@ -1,443 +1,443 @@
+ # steinitzprimedegree (double-indexed dict:
+ #                    stpd[r][k] will be integer corresponding to Steinitz number for polynomial fr_k)
+ # standardextensions (Dict{Integer, FField})
+ # NOTE I think a blanket overwrite of getproperty/setproperty! has potential for conflict with other packages
+ #       that also want to use custom properties...
+ # TODO We should give some type checking on these bonus properties
+ # TODO and also redefine propertynames
+ _propertysymbols = [:is_standardfinitefield, :is_standardprimefield, :primitivepowersintowerbasis, :towerbasis, :steinitzprimedegree, :standardextensions]
+ function Base.setproperty!(F::FinField, v::Symbol, c)
+     if v == :primitivepowersintowerbasis
+         if !isdefined(F, :__attrs)
+             setfield!(F, :__attrs, Dict{Symbol, Any}(v => c))
+             push!(getfield(F, :__attrs), :towerbasis => inv(c))
+         else
+             push!(getfield(F, :__attrs), v =>c)
+             push!(getfield(F, :__attrs), :towerbasis => inv(c))
+         end
+     elseif v == :towerbasis
+         error()
+     elseif v in _propertysymbols
+         if !isdefined(F, :__attrs)
+             setfield!(F, :__attrs, Dict{Symbol, Any}(v => c))
+         else
+             push!(getfield(F, :__attrs), v =>c)
+         end
+     else
+         setfield!(F, v, c)
+     end
+ end
+
+ function Base.getproperty(F::FinField, v::Symbol)
+     if v == :is_standardfinitefield
+         if !isdefined(F, :__attrs) || !haskey(getfield(F, :__attrs), :is_standardfinitefield)
+             return false
+         else
+             return get(getfield(F, :__attrs), :is_standardfinitefield, false)
+         end
+     elseif v == :is_standardprimefield
+         if !isdefined(F, :__attrs) || !haskey(getfield(F, :__attrs), :is_standardprimefield)
+             return false
+         else
+             return get(getfield(F, :__attrs), :is_standardprimefield, false)
+         end
+     elseif v in _propertysymbols
+         if !isdefined(F, :__attrs)
+             error()
+         else
+             return get(getfield(F, :__attrs), v, nothing)
+         end
+     else
+         return getfield(F, v)
+     end
+ end
+
+
+ # TODO: Lübeck speeds this up by caching triples [q,m,a] resulting from this
+ function standardaffineshift(q::ZZRingElem, i)
+     m = div(4*q, 5)
+     while gcd(m,q) != 1
+         m -= 1
+     end
+     a = div(2*q, 3)
+     return mod((m*i + a), q)
+ end
+
+ # Given a field F and Steinitz number n, give the corresponding field element.
+ # we REQUIRE that F is a standard finite field TODO: using @assert?
+ function elementfromsteinitznumber(F::FinField, n)
+     p = characteristic(F)
+     q = order(F)
+     if n < 0 || n > q
+         error("We need to have 0 <= n <= q")
+     end
+     if n == 0
+         return zero(F)
+     elseif q == p
+         return F(n)
+     else
+         vectorrep = GF(p).(digits(n, base = Int(p)))
```

Comment:
Do we really need to cast to `GF(p)` here, can't we form the linear combination below using ad-hoc conversion?